test/hpr_packet_reporter_SUITE.erl

```diff
@@ -0,0 +1,187 @@
+    init_per_testcase/2,
+    end_per_testcase/2
+]).
+
+-export([
+    upload_report_test/1,
+    upload_window_test/1
+]).
+
+-include("hpr.hrl").
+-include("../src/grpc/autogen/server/packet_router_pb.hrl").
+-include("../src/grpc/autogen/server/config_pb.hrl").
+
+-include_lib("common_test/include/ct.hrl").
+-include_lib("eunit/include/eunit.hrl").
+
+-record(state, {
+    aws_client :: aws_client:aws_client(),
+    write_dir :: string(),
+    file_path :: string(),
+    max_file_size :: non_neg_integer(),
+    report_interval :: timer:tref() | undefined,
+    interval_duration :: non_neg_integer() | undefined,
+    upload_window :: non_neg_integer(),
+    upload_window_start_time :: non_neg_integer(),
+    bucket :: binary()
+}).
+
+%%--------------------------------------------------------------------
+%% COMMON TEST CALLBACK FUNCTIONS
+%%--------------------------------------------------------------------
+
+%%--------------------------------------------------------------------
+%% @public
+%% @doc
+%%   Running tests for this suite
+%% @end
+%%--------------------------------------------------------------------
+all() ->
+    [
+        upload_report_test,
+        upload_window_test
+    ].
+
+%%--------------------------------------------------------------------
+%% TEST CASE SETUP
+%%--------------------------------------------------------------------
+init_per_testcase(TestCase, Config) ->
+    Config1 = test_utils:init_per_testcase(TestCase, Config),
+    meck:new(aws_s3, [passthrough]),
+    meck:new(hpr_packet_reporter, [passthrough]),
+
+    Config1.
+
+%%--------------------------------------------------------------------
+%% TEST CASE TEARDOWN
+%%--------------------------------------------------------------------
+end_per_testcase(TestCase, Config) ->
+    ets:delete_all_objects(hpr_packet_report_ets),
+    meck:unload(aws_s3),
+    meck:unload(hpr_packet_reporter),
+
+    test_utils:end_per_testcase(TestCase, Config).
+
+%%--------------------------------------------------------------------
+%% TEST CASES
+%%--------------------------------------------------------------------
+
+upload_report_test(_Config) ->
+    State = sys:get_state(hpr_packet_reporter),
+    #state{file_path = FilePath, aws_client = AWSClient, bucket = Bucket} = State,
+
+    Packet = test_utils:join_packet_up(#{}),
+    Packet2 = test_utils:uplink_packet_up(#{}),
+    Route = test_utils:packet_route(#{}),
+
+    %% Reported packets are encoded and saved to ETS
+    hpr_packet_reporter:report_packet(Packet, Route),
+    hpr_packet_reporter:report_packet(Packet2, Route),
+
+    ETSValues = ets:tab2list(hpr_packet_report_ets),
+    ?assertEqual(2, length(ETSValues)),
+    [P1, P2] = lists:map(
+        fun({_, P}) ->
+            P
+        end,
+        ETSValues
+    ),
+    verify_packet(Packet, Route, P1),
+    verify_packet(Packet2, Route, P2),
+
+    %% Encoded packets are written to a tmp write file
+    hpr_packet_reporter:handle_cast(write, State),
```

Comment:
Use the public API instead? `hpr_packet_report:write_packets/0`? Same for upload?

Reply:
Ah right, they are casts. I hadn't considered that. There's no easy way to deal with that except to add an explicit yield or short sleep. Given that. I think it's fine to call handle_cast directly here.

Reply:
You could do a test only export of the writing and uploading functions. And directly call those.